### ISSUE TITLE
Use trained CDSSM model to predict on another dataset

### ISSUE THREAD

**Issue by logicmd** [NTSC-Community/MatchZoo#344]
Hi 
I used MatchZoo to train a CDSSM model `cdssm_classify.weights.1000`. However I don't know how to use the trained model to predict on another dataset input.csv.

The format of `input.csv` is below

> 
> continue read the king of cards\tresume watching house of cards
> ...
> 

I saw documents say that

> python matchzoo/main.py --phase predict --model_file examples/mymodel/config/cdssm_classify.config

However, I don't know how to generate `relation_file` in `prediction` phase. I presume the model is based ids(triletter_dict.txt, word_dict.txt, word_triletter_map.txt) generated in training phase. If I rerun `test_preparation_for_classify.py` on `input.csv`, I guess I will have a different ids(triletter_dict.txt, word_dict.txt, word_triletter_map.txt) which would cause issues when running existing models.

```
"predict": {
        "input_type": "Triletter_PointGenerator",
        "dtype": "cdssm",
        "phase": "PREDICT",
        "batch_size": 128,
        "relation_file": "/home/ec2-user/oneshot_cleaned/relation_test.txt"
    }
```

**Comment by thiziri** [comment]
Hi @logicmd ,
In fact in order to have the test set processed, you have to merge it with your initial dataset to process and generate the whole vocabulary and ids at once, here is the process:
`dataset = test_data U train_data
 preprocess(dataset)  # make sure to use the function: run_with_train_valid_test_corpus
 --> all data pre-processed`
All you need is to create the "train", "test" and "valid" dataset files then run the preparation processing using [run_with_train_valid_test_corpus](run_with_train_valid_test_corpus)
Good luck :)

**Comment by logicmd** [comment]
Hi @thiziri ,

Thanks for the reply. I understand your point. You suggested I add the dataset which I want to predict into train and test dataset and go through the preprocessing to generate the id hash map.

While it is feasible, it would be still better to decouple  preprocessing from training. Your proposed approach would force the model to be retrained if I want to predict on another dataset which would be unrealistic in production environment, I can't guarantee two exactly same model even if I use the same data and hyper-parameter to train the model.  

Could we work on decouple preprocessing from training? It would be nice if there is a separate interface to id-hash test data by leveraging existing id-hash map to avoid model retraining.

Thanks

**Comment by ZeonTrevor** [comment]
Hi @logicmd, do have a look at #236, I think that's relevant to your query.

**Comment by bwanglzu** [comment]
@logicmd @thiziri @ZeonTrevor Thanks for your reply, take a look at matchzoo 2.0 api [here](http://nbviewer.jupyter.org/github/faneshion/MatchZoo/blob/2.0/notebooks/wikiqa/dssm.ipynb).

**Comment by logicmd** [comment]
@thiziri , it would be better that MatchZoo could support missing value. As a workaround, what you could do is to skip the tri-letter which is not in existing id map.
Please let me know if that works or not.

**Comment by logicmd** [comment]
Thank you for letting me know the plans and update. @bwanglzu

**Comment by logicmd** [comment]
Hi @bwanglzu , are we still on track about 2.0 release?

**Comment by logicmd** [comment]
here is the output I got from model output
anyone know its schema?

```

T2446717        Q0      T59     0       0.177585        cdssm   0
T2446717        Q0      T730    1       0.177568        cdssm   0
T2446717        Q0      T2135   2       0.177568        cdssm   0
T2446718        Q0      T5572   0       0.178070        cdssm   0
T2446718        Q0      T1062   1       0.177568        cdssm   0
T2446718        Q0      T266    2       0.177568        cdssm   0
T2446718        Q0      T43     3       0.177568        cdssm   0
T2446718        Q0      T3615   4       0.177568        cdssm   0
T2446719        Q0      T470    0       0.177614        cdssm   1
```

**Comment by bwanglzu** [comment]
@logicmd the plan is to finalize dssm, cdssm and arc1 before oct 1. See pull request, it's done waiting for review.

**Comment by logicmd** [comment]
Thank you @bwanglzu

**Comment by logicmd** [comment]
@bwanglzu do we have 2.0 in master branch? Or do you still recommend 2.0 branch to use 2.0?

**Comment by logicmd** [comment]
yes, as we discussed, 2.0 would decouple training and evaluation using CDSSM model which is exactly what I need.

**Comment by bwanglzu** [comment]
@logicmd By this weekend I'll write a tutorial on "train a CDSSM model using matchzoo 2.0", currently @ZizhenWang is working on CDSSM integration test, once it's merged it should be done done.

The only possible issue is that you can only use Keras metrics at the moment since the MatchZoo customized metrics (see metrics.py undetr 2.0 branch) need refactoring.
